### DISCUSSION TITLE
Database install error on PostgreSQL (CockroachDB)

### DISCUSSION THREAD

**Discussion by adamjgrant** [directus/directus#9268] [Archive]
*Update: This appears to be a quirk with cockroachDB's implementation of postgres. I'll post my findings here while I investigate a workaround.*

### Preflight Checklist

- [X] I have completed all [Troubleshooting Steps](https://docs.directus.io/getting-started/support/#troubleshooting-steps).
- [X] I'm on [the latest version of Directus](https://github.com/directus/directus/releases).
- [X] There's [no other issue](https://github.com/directus/directus/issues) that already describes my problem.

### Describe the Bug

Database install fails on first table install attempt with error, `alter table "directus_collections" add constraint "directus_collections_pkey" primary key ("collection") - cannot use nullable column "collection" in primary key`

To be sure, this is a fresh instance of cockroachdb and database.

### To Reproduce

Installing Directus as prescribed using Ubuntu 20.04.1 LTS and a remote cockroachdb (postgreSQL) install.

### Errors Shown

~~~
Something went wrong while seeding the database:

[42P15] alter table "directus_collections" add constraint "directus_collections_pkey" primary key ("collection") - cannot use nullable column "collection" in primary key

Please try again
~~~

### What version of Directus are you using?

(9.0.0-rc.100) Friendly request to guide users how to check this. The guide just links to the release in github which only gets you halfway there. Lots of google searching and one hour later allowed me to finally yield this answer.

### What version of Node.js are you using?

v14.15.4

### What database are you using?

Postgres 12.5 (cockroachdb)

### What browser are you using?

N/A

### What operating system are you using?

Ubuntu 20.04.1 LTS

### How are you deploying Directus?

npm install on Ubuntu

**Comment by adamjgrant** [comment]
# Reproducibility
I've created an identical and empty, unused, throwaway db for reproduction which I'd be happy to share the credentials for (not in this GH issue). Please advise if desired.

Also looks like [native support in knex is on its way](https://github.com/knex/knex/issues/2002#issuecomment-944593235).

---

[Monday November 1, 2021] Simplifying this huge comment...

I force upgraded directus to use the latest release of knex, `v0.95.12` which explicitly includes a new dialect for CDB
However, the results are the same.

~~~
Something went wrong while seeding the database:

[42P15] alter table "directus_collections" add constraint "directus_collections_pkey" primary key ("collection") - cannot use nullable column "collection" in primary key

Please try again

? Database Host: (127.0.0.1) 
~~~

Note, to get this to run it currently requires changing `/node_modules/knex/lib/dialects/postgres/index.js` Line 130 which expects to parse the PG version number from a string beginning with "Postgres". CockroachDB returns "pgsl (Postgres)". I changed this to just `return "12.5";`

Not a solution, but I'm trying to get to the bottom of what is left to do to get directus compatible with CDB. I've also alerted Knex about this issue.

**Reply by azrikahar** [reply]
Thanks for keeping us updated! Personally don't have actual experience using CDB so these updates/investigations are very helpful.

**Reply by rijkvanzanten** [reply]
> I force upgraded directus to use the latest release of knex, v0.95.12 which explicitly includes a new dialect for CDB

This doesn't mean Directus uses it though. In Knex, you have to tell it what dialect to use explicitly, and at current time, Directus doesn't support passing the cockroach configuration to knex. Next to this, Directus implements a couple helper functions and functionality that Knex doesn't support natively, so we'll have to create those for cockroach as well before we can finalize supporting cockroach for real 🙂

**Reply by adamjgrant** [reply]
Thanks. That's helpful information. 

My goal is to be your first user running this set up end to end and share any knowledge back to induct Directus onto CDB.

With their new serverless model starting at $0,  I think this integration could be a big boost for directus adoption. Imagine keeping directus files in Lambda and using this as a db. You could spin up personal projects for pennies or grow startups from tiny budgets.

**Reply by rijkvanzanten** [reply]
Absolutely!! We'll have to check/update a couple points, like [the functions here](https://github.com/directus/directus/tree/main/api/src/database/functions/dialects), the helpers here https://github.com/directus/directus/tree/main/api/src/database/helpers and we'll have to add [the dialect here](https://github.com/knex/knex-schema-inspector/tree/master/lib/dialects). All those places are highly postgres specific, so we'll have to ensure compatibility with the queries we make in order to ensure cockroach will be supported properly.

**Comment by azrikahar** [comment]
Although CockroachDB does provide support to be compatible to PostgreSQL, technically it is still not within the current list of supported database. This is why I opted to move this to discussion at this point in time. However we are indeed keen on it especially with the native support in Knex as you mentioned!

---

> (9.0.0-rc.100) Friendly request to guide users how to check this. The guide just links to the release in github which only gets you halfway there. Lots of google searching and one hour later allowed me to finally yield this answer.

https://user-images.githubusercontent.com/42867097/139570605-e7d1fec7-7bbd-47c3-9177-631c29b5a398.mp4

One way to check this via the App would be go to the Settings, and it should show the version at the end of the navigation list. Clicking the Report Bug link will also bring to the GH issues page with it autofilled for you. Feel free to let us know if there's improvements we can make to facilitate this in an easier way.

> Somewhere to unblock all these migrations?

Probably others can give better insights in this matter, but we do try to avoid slipping in too many database vendor specific code (as that can be quite difficult to manage over time). However this definitely a big step in the venture of officially adding CockroachDB support for Directus so really appreciate you paving the way! 👍 let's see what are the thoughts from the rest of team.

**Reply by rijkvanzanten** [reply]
Yeah I'd rather look into officially supporting CockroachDB through the official adapters in Knex than trying to hack around our current Postgres implementation to support Cockroach as well 👍🏻

**Reply by adamjgrant** [reply]
Diagnosis steps ≠ suggestion for a solution.

**Reply by adamjgrant** [reply]
@azrikahar 

> One way to check this via the App would be go to the Settings

Thank you for the video but this issue is regarding the inability to get Directus installed and running, so how does one "go to the settings"?

**Reply by adamjgrant** [reply]
@azrikahar 

> but we do try to avoid slipping in too many database vendor specific code (as that can be quite difficult to manage over time)

I don't think you'd want to, this is a quirk in one of your modules that implements the dialect for CDB (knex).

**Reply by azrikahar** [reply]
> Diagnosis steps ≠ suggestion for a solution.

This is true. On hindsight the way I worded it felt like you're suggesting it, but it was definitely not my intention so hope you mind it 👍

> Thank you for the video but this issue is regarding the inability to get Directus installed and running, so how does one "go to the settings"?

Thanks for clarifying! Hmm... just to confirm, I assume this is similar/roughly what you saw right?

![WindowsTerminal_BmeH87Mqh9](https://user-images.githubusercontent.com/42867097/139801409-0f151c57-9e80-4f37-b483-15fdedb7f6a5.png)

In that case there's definitely no visibility of the version here, even though the cli does offer the typical `-v`/`--version` flag: https://github.com/directus/directus/blob/ea4393bda174e3d852880e7f85df350514fc5547/api/src/cli/index.ts#L28

Perhaps we can do something similar to what `create-nuxt-app` (likely others as well, this is just what came in mind) does over here: https://github.com/nuxt/create-nuxt-app/blob/cdd25976cfd24a799e3af5fef24a91f2fc1061ba/packages/create-nuxt-app/lib/cli.js#L39 to at least display the version. @rijkvanzanten thoughts on this?

**Reply by adamjgrant** [reply]
Honestly, I think having the value visible in package.json is sufficient but the chain of npm/x commands that get you to having the monorepo obfuscates how to find it a bit. 

I was thinking this could just be a sentence in the steps you have listed for opening a PR.

**Comment by adamjgrant** [comment]
Following on [this](https://github.com/directus/directus/discussions/9268#discussioncomment-1576997), dumping my breadcrumbs on steps so far here. No action needed. I'm going to just keep saving and editing this as I go.

* Fresh install of node.js droplet on digitalocean (not using the one-click directus offering)
* Following [steps here](https://docs.directus.io/getting-started/installation/manual/), `mkdir` and `cd` into `/var/www/hub`, then `npm install --save directus`. Updating package.json to autostart per instructions. Env file stuff with CockroachDB (CDB) creds.

~~~
DB_CLIENT="pg"
DB_HOST="XXXXXX.cockroachlabs.cloud"
DB_PORT=26257
DB_DATABASE="XXXXX.directus"
DB_USER="adam"
DB_PASSWORD="XXXX"
~~~

* cd `/var/www/hub/node_modules/ && npm install "git://github.com/knex/knex.git#master" --save` to use the latest master branch version of knex. Note this is pre-release.
* Just to be sure...

    ~~~
    grep -E "\"version\"" /var/www/hub/node_modules/knex/package.json 
      "version": "1.0.0",
    ~~~

* Rebooting for the hell of it.
* `cd /var/www/hub && npx directus init` (even though I did the env file stuff already)

🐰 **Rabbit hole**: As it stands with knex's master branch today, users won't get past this step no matter what we change in directus.

__I have [an incredibly sh**y PR](https://github.com/knex/knex/pull/4785) I opened to address this but will need to__

[Wednesday November 3, 2021 12:17 PM] Nope. They should have this in 1.0.0 already but I haven't done anything to get directus to use that dialect.

(1) Come up with a better regex than I proposed. (Quick)
(2) Fully set up/use code spaces to write a test to go with the PR per their contrib guidelines (Long).

~~~
? Choose your database client PostgreSQL / Redshift
? Database Host: XXXX.cockroachlabs.cloud
? Port: 26257
? Database Name: XXXX.directus
? Database User: adam
? Database Password: **********
? Enable SSL: Yes
/var/www/hub/node_modules/directus/node_modules/knex/lib/dialects/postgres/index.js:130
    return /^PostgreSQL (.*?)( |$)/.exec(versionString)[1];
~~~

**Comment by adamjgrant** [comment]
Trying to do this again but with a fork of https://github.com/directus/directus

Starting with installing 1.0.0 of knex, I'm in a rabbit hole of errors that I don't have various things installed.

Checked the readme.md but it looks like it's more for consumers than contributors on the repo.
No contributing.md. Checked [wiki](https://github.com/directus/directus/wiki).

Tried npm install, npm run build, various things like lerna, rimraf need to be installed.

Is there a single point of entry somewhere where I can bootstrap the repo?

@rijkvanzanten @azrikahar

**Reply by rijkvanzanten** [reply]
Have you seen https://docs.directus.io/contributing/running-locally/#running-locally?

**Reply by danilopolani** [reply]
Also be sure you have at least Node16 and Npm8 to build everything correctly, otherwise a lot of builds will fail (personal experience lol)

**Reply by adamjgrant** [reply]
Cheers. I did miss that page. Thanks.

**Comment by wodka** [comment]
As I'd like to have cockroach working as well I "hotfixed" all entries till I could finish the installation.

Next I did create a role ` yarn directus roles create --role admin --admin ` and then a user `yarn directus users create --email michael@local --password test --role fd8ccd18-22fe-4ef9-9ceb-03a395836500`

The Login itself worked but now I am stuck at a new Error:

![image](https://user-images.githubusercontent.com/385731/143647281-10b0e290-ba17-4227-aabd-0446918a6698.png)

<details>
  <summary>`relation "active_range_feeds" does not exist`</summary>

```
22:05:19 ✨ Server started at http://localhost:8055
22:05:28 ✨ request completed GET 304 /admin/content 5ms
22:05:28 ✨ request completed GET 304 /admin/index.b5d16ba8.js 3ms
22:05:28 ✨ request completed GET 304 /admin/assets/vendor.c46e17d6.js 2ms
22:05:28 ✨ request completed GET 304 /admin/assets/vendor.fe7856f0.css 1ms
22:05:28 ✨ request completed GET 304 /admin/assets/use-sync.864aa3cf.js 1ms
22:05:28 ✨ request completed GET 304 /admin/assets/index.972d7652.css
22:05:28 ✨ request completed GET 200 /extensions/panels/index.js 31ms
22:05:28 ✨ request completed GET 200 /extensions/layouts/index.js 31ms
22:05:28 ✨ request completed GET 200 /extensions/displays/index.js 34ms
22:05:28 ✨ request completed GET 200 /extensions/interfaces/index.js 39ms
22:05:28 ✨ request completed GET 200 /extensions/modules/index.js 35ms
22:05:28 ✨ request completed POST 200 /auth/refresh 51ms
22:05:28 ✨ request completed GET 304 /admin/img/icons/android-chrome-512x512.png 1ms
22:05:28 ✨ request completed GET 200 /server/info?limit=-1 25ms
22:05:28 ✨ request completed GET 304 /admin/img/icons/android-chrome-192x192.png 3ms
22:05:28 ✨ request completed GET 304 /admin/manifest.webmanifest 1ms
22:05:29 ✨ request completed GET 304 /admin/img/icons/safari-pinned-tab.svg 1ms
22:05:29 ✨ request completed GET 304 /admin/img/icons/android-chrome-192x192.png 7ms
22:05:29 ✨ request completed GET 200 /users/me?fields=*,avatar.id,role.* 49ms
22:05:29 ✨ request completed GET 304 /permissions?limit=-1&filter=%7B%22role%22:%7B%22_eq%22:%22fd8ccd18-22fe-4ef9-9ceb-03a395836500%22%7D%7D 27ms
22:05:29 🚨 error select "c"."column_name", "c"."table_name", "c"."data_type", "c"."column_default", "c"."character_maximum_length", "c"."is_generated", "c"."is_nullable", "c"."numeric_precision", "c"."numeric_scale", "c"."table_schema", "c"."is_identity", "c"."generation_expression", pg_get_serial_sequence(quote_ident(c.table_name), c.column_name) as serial, pg_catalog.col_description(pg_class.oid, c.ordinal_position:: int) as column_comment, COALESCE(pg.indisunique, false) as is_unique, COALESCE(pg.indisprimary, false) as is_primary, "ffk"."foreign_key_schema", "ffk"."foreign_key_table", "ffk"."foreign_key_column" from information_schema.columns c 
        LEFT JOIN pg_catalog.pg_class
          ON pg_catalog.pg_class.oid = quote_ident(c.table_name):: regclass:: oid
          AND pg_catalog.pg_class.relname = c.table_name
       
        LEFT JOIN LATERAL (
          SELECT
            pg_index.indisprimary,
            pg_index.indisunique
          FROM pg_index
          JOIN pg_attribute
            ON pg_attribute.attrelid = pg_index.indrelid
            AND pg_attribute.attnum = any(pg_index.indkey)
          WHERE pg_index.indrelid = quote_ident(c.table_name)::regclass
          AND pg_attribute.attname = c.column_name
          LIMIT 1
        ) pg ON true
       
        LEFT JOIN LATERAL (
          SELECT
            k2.table_schema AS foreign_key_schema,
            k2.table_name AS foreign_key_table,
            k2.column_name AS foreign_key_column
          FROM
            information_schema.key_column_usage k1
            JOIN information_schema.referential_constraints fk using (
              constraint_schema, constraint_name
            )
            JOIN information_schema.key_column_usage k2
              ON k2.constraint_schema = fk.unique_constraint_schema
              AND k2.constraint_name = fk.unique_constraint_name
              AND k2.ordinal_position = k1.position_in_unique_constraint
            WHERE k1.table_name = c.table_name
            AND k1.column_name = c.column_name
        ) ffk ON TRUE
       where "c"."table_schema" in ($1) order by "c"."table_name" asc, "c"."ordinal_position" asc - relation "active_range_feeds" does not exist 
error: select "c"."column_name", "c"."table_name", "c"."data_type", "c"."column_default", "c"."character_maximum_length", "c"."is_generated", "c"."is_nullable", "c"."numeric_precision", "c"."numeric_scale", "c"."table_schema", "c"."is_identity", "c"."generation_expression", pg_get_serial_sequence(quote_ident(c.table_name), c.column_name) as serial, pg_catalog.col_description(pg_class.oid, c.ordinal_position:: int) as column_comment, COALESCE(pg.indisunique, false) as is_unique, COALESCE(pg.indisprimary, false) as is_primary, "ffk"."foreign_key_schema", "ffk"."foreign_key_table", "ffk"."foreign_key_column" from information_schema.columns c 
        LEFT JOIN pg_catalog.pg_class
          ON pg_catalog.pg_class.oid = quote_ident(c.table_name):: regclass:: oid
          AND pg_catalog.pg_class.relname = c.table_name
       
        LEFT JOIN LATERAL (
          SELECT
            pg_index.indisprimary,
            pg_index.indisunique
          FROM pg_index
          JOIN pg_attribute
            ON pg_attribute.attrelid = pg_index.indrelid
            AND pg_attribute.attnum = any(pg_index.indkey)
          WHERE pg_index.indrelid = quote_ident(c.table_name)::regclass
          AND pg_attribute.attname = c.column_name
          LIMIT 1
        ) pg ON true
       
        LEFT JOIN LATERAL (
          SELECT
            k2.table_schema AS foreign_key_schema,
            k2.table_name AS foreign_key_table,
            k2.column_name AS foreign_key_column
          FROM
            information_schema.key_column_usage k1
            JOIN information_schema.referential_constraints fk using (
              constraint_schema, constraint_name
            )
            JOIN information_schema.key_column_usage k2
              ON k2.constraint_schema = fk.unique_constraint_schema
              AND k2.constraint_name = fk.unique_constraint_name
              AND k2.ordinal_position = k1.position_in_unique_constraint
            WHERE k1.table_name = c.table_name
            AND k1.column_name = c.column_name
        ) ffk ON TRUE
       where "c"."table_schema" in ($1) order by "c"."table_name" asc, "c"."ordinal_position" asc - relation "active_range_feeds" does not exist
    at Parser.parseErrorMessage (/directus/node_modules/pg-protocol/dist/parser.js:287:98)
    at Parser.handlePacket (/directus/node_modules/pg-protocol/dist/parser.js:126:29)
    at Parser.parse (/directus/node_modules/pg-protocol/dist/parser.js:39:38)
    at Socket.<anonymous> (/directus/node_modules/pg-protocol/dist/index.js:11:42)
    at Socket.emit (node:events:390:28)
    at addChunk (node:internal/streams/readable:315:12)
    at readableAddChunk (node:internal/streams/readable:289:9)
    at Socket.Readable.push (node:internal/streams/readable:228:10)
    at TCP.onStreamRead (node:internal/stream_base_commons:199:23)
22:05:29 ✨ request errored GET 500 /fields?limit=-1 125ms
22:05:29 ✨ request completed GET 304 /collections?limit=-1 125ms
22:05:29 ✨ request completed GET 304 /admin/assets/Inter-Medium.a4e1e7e6.woff2 1ms
22:05:29 ✨ request completed GET 304 /admin/assets/Inter-Bold.f0f54a06.woff2 1ms
22:05:29 ✨ request completed GET 304 /admin/assets/material-icons-outline.63af6c49.woff2 1ms
22:05:29 ✨ request completed GET 304 /admin/assets/FiraMono-Medium.33e88fa9.woff2 1ms
22:05:29 ✨ request completed GET 304 /presets?filter[user][_eq]=7b3ba763-a942-45b6-988d-dc9e4b515dcd&limit=-1 25ms
22:05:29 ✨ request completed GET 304 /presets?filter[role][_eq]=fd8ccd18-22fe-4ef9-9ceb-03a395836500&filter[user][_null]=true&limit=-1 24ms
22:05:29 ✨ request completed GET 304 /presets?filter[role][_null]=true&filter[user][_null]=true&limit=-1 24ms
22:05:29 ✨ request completed GET 304 /settings 27ms
22:05:29 ✨ request completed GET 200 /server/info?limit=-1 32ms
22:05:29 ✨ request completed GET 304 /admin/img/icons/android-chrome-512x512.png 2ms
22:05:29 ✨ request completed GET 304 /notifications?filter=%7B%22_and%22:[%7B%22recipient%22:%7B%22_eq%22:%227b3ba763-a942-45b6-988d-dc9e4b515dcd%22%7D%7D,%7B%22status%22:%7B%22_eq%22:%22inbox%22%7D%7D]%7D&aggregate=%7B%22count%22:%22id%22%7D 27ms
22:05:29 ✨ request completed GET 304 /dashboards?limit=-1&fields[]=*&fields[]=panels.*&sort[]=name 30ms
22:05:30 ✨ request completed PATCH 200 /users/me/track/page 43ms
22:05:30 ✨ request completed GET 200 /relations?limit=-1 152ms
```
</details>

on a further note it did fail on relation `backward_dependencies` with crdb 21.1, updating to 21.2 shows my current error.

In theory those should be supported: https://www.cockroachlabs.com/docs/v21.2/crdb-internal.html

**Reply by wodka** [reply]
ok, this boils down to https://github.com/knex/knex-schema-inspector not support cockroachdb

**Reply by wodka** [reply]
Fixing the queries in knex-schema (https://github.com/knex/knex-schema-inspector/issues/86) brings another new error:

![image](https://user-images.githubusercontent.com/385731/143657040-627ce716-544d-4d1b-a0ed-1bb89716942b.png)

```
22:45:05 ✨ request completed PATCH 200 /users/me/track/page 23ms
22:45:22 🚨 error create table "test" ("id" serial primary key, "status" varchar(255) not null default 'draft', "user_created" uuid null, "date_created" timestamptz null, "user_updated" uuid null, "date_updated" timestamptz null) - unimplemented: the first schema change statement in a transaction must precede any writes 
error: create table "test" ("id" serial primary key, "status" varchar(255) not null default 'draft', "user_created" uuid null, "date_created" timestamptz null, "user_updated" uuid null, "date_updated" timestamptz null) - unimplemented: the first schema change statement in a transaction must precede any writes
    at Parser.parseErrorMessage (/directus/node_modules/pg-protocol/dist/parser.js:287:98)
    at Parser.handlePacket (/directus/node_modules/pg-protocol/dist/parser.js:126:29)
    at Parser.parse (/directus/node_modules/pg-protocol/dist/parser.js:39:38)
    at Socket.<anonymous> (/directus/node_modules/pg-protocol/dist/index.js:11:42)
    at Socket.emit (node:events:390:28)
    at addChunk (node:internal/streams/readable:315:12)
    at readableAddChunk (node:internal/streams/readable:289:9)
    at Socket.Readable.push (node:internal/streams/readable:228:10)
    at TCP.onStreamRead (node:internal/stream_base_commons:199:23)
22:45:22 ✨ request errored POST 500 /collections 59ms
```

**Reply by rijkvanzanten** [reply]
@wodka Directus was never made to work on CockroachDB 😬 It seems like their Postgres compatibility layer doesn't reach far enough for Directus to just work on it out of the box. That being said, seeing that Knex now natively supports CockroachDB, it shouldn't be _too_ difficult to add in support for 🪳 as well

**Reply by wodka** [reply]
;) so I'm trying to fix everything that is needed to get it to work

Long way to go: https://github.com/directus/directus/pull/10113

**Reply by wodka** [reply]
so far it seems with just the few modifications that everything works (wrapping alter tables in transactions + updating schema inspector)
